File: source/bindings/csharp/recognizer.cs
//
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
//

using System;
using System.Runtime.InteropServices;
using Microsoft.CognitiveServices.Speech.Internal;
using static Microsoft.CognitiveServices.Speech.Internal.SpxExceptionThrower;

namespace Microsoft.CognitiveServices.Speech
{
    /// <summary>
    /// Defines the base class Recognizer which mainly contains common event handlers.
    /// </summary>
    public class Recognizer : IDisposable
    {
        private event EventHandler<SessionEventArgs> _SessionStarted;
        private event EventHandler<SessionEventArgs> _SessionStopped;
        private event EventHandler<RecognitionEventArgs> _SpeechStartDetected;
        private event EventHandler<RecognitionEventArgs> _SpeechEndDetected;

        /// <summary>
        /// Defines event handler for session started event.
        /// </summary>
        public event EventHandler<SessionEventArgs> SessionStarted
        {
            add
            {
                if (this._SessionStarted == null)
                {
                    ThrowIfFail(Internal.Recognizer.recognizer_session_started_set_callback(recoHandle, sessionStartedCallbackDelegate, GCHandle.ToIntPtr(gch)));
                }
                this._SessionStarted += value;
            }
            remove
            {
                this._SessionStarted -= value;
                if (this._SessionStarted == null)
                {
                    LogErrorIfFail(Internal.Recognizer.recognizer_session_started_set_callback(recoHandle, null, IntPtr.Zero));
                }
            }
        }

        /// <summary>
        /// Defines event handler for session stopped event.
        /// </summary>
        public event EventHandler<SessionEventArgs> SessionStopped
        {
            add
            {
                if (this._SessionStopped == null)
                {
                    ThrowIfFail(Internal.Recognizer.recognizer_session_stopped_set_callback(recoHandle, sessionStoppedCallbackDelegate, GCHandle.ToIntPtr(gch)));
                }
                this._SessionStopped += value;
            }
            remove
            {
                this._SessionStopped -= value;
                if (this._SessionStopped == null)
                {
                    LogErrorIfFail(Internal.Recognizer.recognizer_session_stopped_set_callback(recoHandle, null, IntPtr.Zero));
                }
            }
        }

        /// <summary>
        /// Defines event handler for speech start detected event.
        /// </summary>
        public event EventHandler<RecognitionEventArgs> SpeechStartDetected
        {
            add
            {
                if (this._SpeechStartDetected == null)
                {
                    ThrowIfFail(Internal.Recognizer.recognizer_speech_start_detected_set_callback(recoHandle, speechStartDetectedCallbackDelegate, GCHandle.ToIntPtr(gch)));
                }
                this._SpeechStartDetected += value;
            }
            remove
            {
                this._SpeechStartDetected -= value;
                if (this._SpeechStartDetected == null)
                {
                    LogErrorIfFail(Internal.Recognizer.recognizer_speech_start_detected_set_callback(recoHandle, null, IntPtr.Zero));
                }
            }
        }

        /// <summary>
        /// Defines event handler for speech end detected event.
        /// </summary>
        public event EventHandler<RecognitionEventArgs> SpeechEndDetected
        {
            add
            {
                if (this._SpeechEndDetected == null)
                {
                    ThrowIfFail(Internal.Recognizer.recognizer_speech_end_detected_set_callback(recoHandle, speechEndDetectedCallbackDelegate, GCHandle.ToIntPtr(gch)));
                }
                this._SpeechEndDetected += value;
            }
            remove
            {
                this._SpeechEndDetected -= value;
                if (this._SpeechEndDetected == null)
                {
                    LogErrorIfFail(Internal.Recognizer.recognizer_speech_end_detected_set_callback(recoHandle, null, IntPtr.Zero));
                }
            }
        }

        internal Recognizer(InteropSafeHandle recoPtr)
        {
            ThrowIfNull(recoPtr);
            recoHandle = recoPtr;
            gch = GCHandle.Alloc(this, GCHandleType.Weak);
            speechStartDetectedCallbackDelegate = FireEvent_SpeechStartDetected;
            speechEndDetectedCallbackDelegate = FireEvent_SpeechEndDetected;
            sessionStartedCallbackDelegate = FireEvent_SetSessionStarted;
            sessionStoppedCallbackDelegate = FireEvent_SetSessionStopped;
        }

        ~Recognizer()
        {
            isDisposing = true;
            Dispose(false);
        }

        /// <summary>
        /// Dispose of associated resources.
        /// </summary>
        public void Dispose()
        {
            try
            {
                isDisposing = true;
                lock (recognizerLock)
                {
                    if (activeAsyncRecognitionCounter != 0)
                    {
                        throw new InvalidOperationException("Cannot dispose a recognizer while async recognition is running. Await async recognitions to avoid unexpected disposals.");
                    }
                }
            }
            finally
            {
                Dispose(true);
                GC.SuppressFinalize(this);
            }
        }

        /// <summary>
        /// This method performs cleanup of resources.
        /// The Boolean parameter <paramref name="disposing"/> indicates whether the method is called from <see cref="IDisposable.Dispose"/> (if <paramref name="disposing"/> is true) or from the finalizer (if <paramref name="disposing"/> is false).
        /// Derived classes should override this method to dispose resource if needed.
        /// </summary>
        /// <param name="disposing">Flag to request disposal.</param>
        protected virtual void Dispose(bool disposing)
        {
            if (disposed)
            {
                return;
            }

            speechStartDetectedCallbackDelegate = null;
            speechEndDetectedCallbackDelegate = null;
            sessionStartedCallbackDelegate = null;
            sessionStoppedCallbackDelegate = null;
            if (gch.IsAllocated)
            {
                gch.Free();
            }
            disposed = true;
        }

        internal InteropSafeHandle recoHandle;
        private IntPtr asyncStartContinuousHandle = IntPtr.Zero;
        private IntPtr asyncStopContinuousHandle = IntPtr.Zero;
        private IntPtr asyncStartKeywordHandle = IntPtr.Zero;
        private IntPtr asyncStopKeywordHandle = IntPtr.Zero;
        private CallbackFunctionDelegate speechStartDetectedCallbackDelegate;
        private CallbackFunctionDelegate speechEndDetectedCallbackDelegate;
        private CallbackFunctionDelegate sessionStartedCallbackDelegate;
        private CallbackFunctionDelegate sessionStoppedCallbackDelegate;

        /// <summary>
        /// GC handle for callbacks for context.
        /// </summary>
        protected GCHandle gch;

        /// <summary>
        /// disposed is a flag used to indicate if object is disposed.
        /// </summary>
        protected volatile bool disposed = false;

        /// <summary>
        /// isDisposing is a flag used to indicate if object is being disposed.
        /// </summary>
        protected volatile bool isDisposing = false;

        /// <summary>
        /// recognizerLock is used to synchronize access to objects member variables from multiple threads
        /// </summary>
        protected object recognizerLock = new object();

        private int activeAsyncRecognitionCounter = 0;

        /// <summary>
        /// Define a private methods which raise a C# event when a corresponding callback is invoked from the native layer.
        /// </summary>
        ///
        [MonoPInvokeCallback(typeof(CallbackFunctionDelegate))]
        private static void FireEvent_SetSessionStarted(IntPtr hreco, IntPtr hevent, IntPtr pvContext)
        {
            try
            {
                var recognizer = InteropSafeHandle.GetObjectFromWeakHandle<Recognizer>(pvContext);
                if (recognizer == null || recognizer.isDisposing)
                {
                    return;
                }
                var resultEventArg = new SessionEventArgs(hevent);
                recognizer._SessionStarted?.Invoke(recognizer, resultEventArg);
            }
            catch (InvalidOperationException)
            {
                LogError(SpxError.InvalidHandle);
            }
        }

        [MonoPInvokeCallback(typeof(CallbackFunctionDelegate))]
        private static void FireEvent_SetSessionStopped(IntPtr hreco, IntPtr hevent, IntPtr pvContext)
        {
            try
            {
                var recognizer = InteropSafeHandle.GetObjectFromWeakHandle<Recognizer>(pvContext);
                if (recognizer == null || recognizer.isDisposing)
                {
                    return;
                }
                var resultEventArg = new SessionEventArgs(hevent);
                recognizer._SessionStopped?.Invoke(recognizer, resultEventArg);
            }
            catch (InvalidOperationException)
            {
                LogError(SpxError.InvalidHandle);
            }
        }

        [MonoPInvokeCallback(typeof(CallbackFunctionDelegate))]
        private static void FireEvent_SpeechStartDetected(IntPtr hreco, IntPtr hevent, IntPtr pvContext)
        {
            try
            {
                var recognizer = InteropSafeHandle.GetObjectFromWeakHandle<Recognizer>(pvContext);
                if (recognizer == null || recognizer.isDisposing)
                {
                    return;
                }
                var resultEventArg = new RecognitionEventArgs(hevent);
                recognizer._SpeechStartDetected?.Invoke(recognizer, resultEventArg);
            }
            catch (InvalidOperationException)
            {
                LogError(SpxError.InvalidHandle);
            }
        }

        [MonoPInvokeCallback(typeof(CallbackFunctionDelegate))]
        private static void FireEvent_SpeechEndDetected(IntPtr hreco, IntPtr hevent, IntPtr pvContext)
        {
            try
            {
                var recognizer = InteropSafeHandle.GetObjectFromWeakHandle<Recognizer>(pvContext);
                if (recognizer == null || recognizer.isDisposing)
                {
                    return;
                }
                var resultEventArg = new RecognitionEventArgs(hevent);
                recognizer._SpeechEndDetected?.Invoke(recognizer, resultEventArg);
            }
            catch (InvalidOperationException)
            {
                LogError(SpxError.InvalidHandle);
            }
        }

        /// <summary>
        /// This methods checks if a recognizer is disposed before performing async recognition action.
        /// The Action parameter <paramref name="recoImplAction"/> can be any internal async recognition method of Speech, Translation and Intent Recognizer.
        /// The method is called from all async recognition methods (e.g. <see cref="SpeechRecognizer.StartContinuousRecognitionAsync"/>).
        /// ObjectDisposedException will be thrown and the action will not be performed if its recognizer is not available anymore.
        /// The purpose of this method is to prevent possible race condition if async recognitions are not awaited.
        /// </summary>
        /// <param name="recoImplAction">Actual implementation.</param>
        protected void DoAsyncRecognitionAction(Action recoImplAction)
        {
            lock (recognizerLock)
            {
                activeAsyncRecognitionCounter++;
            }
            if (disposed || isDisposing)
            {
                throw new ObjectDisposedException(this.GetType().Name);
            }
            try
            {
                recoImplAction();
            }
            finally
            {
                lock (recognizerLock)
                {
                    activeAsyncRecognitionCounter--;
                }
            }
        }

        internal IntPtr RecognizeOnce()
        {
            IntPtr hresult = IntPtr.Zero;
            ThrowIfNull(recoHandle, "Invalid recognizer handle");
            ThrowIfFail(Internal.Recognizer.recognizer_recognize_once(recoHandle, out hresult));
            return hresult;
        }

        internal void StartContinuousRecognition()
        {
            if (asyncStartContinuousHandle != IntPtr.Zero)
            {
                ThrowIfFail(Internal.Recognizer.recognizer_async_handle_release(asyncStartContinuousHandle));
            }
            ThrowIfNull(recoHandle, "Invalid recognizer handle");
            ThrowIfFail(Internal.Recognizer.recognizer_start_continuous_recognition_async(recoHandle, out asyncStartContinuousHandle));
            ThrowIfFail(Internal.Recognizer.recognizer_start_continuous_recognition_async_wait_for(asyncStartContinuousHandle, UInt32.MaxValue));
            ThrowIfFail(Internal.Recognizer.recognizer_async_handle_release(asyncStartContinuousHandle));
            asyncStartContinuousHandle = IntPtr.Zero;
        }

        internal void StopContinuousRecognition()
        {
            if (asyncStopContinuousHandle != IntPtr.Zero)
            {
                ThrowIfFail(Internal.Recognizer.recognizer_async_handle_release(asyncStopContinuousHandle));
            }
            ThrowIfNull(recoHandle, "Invalid recognizer handle");
            ThrowIfFail(Internal.Recognizer.recognizer_stop_continuous_recognition_async(recoHandle, out asyncStopContinuousHandle));
            ThrowIfFail(Internal.Recognizer.recognizer_stop_continuous_recognition_async_wait_for(asyncStopContinuousHandle, UInt32.MaxValue));
            ThrowIfFail(Internal.Recognizer.recognizer_async_handle_release(asyncStopContinuousHandle));
            asyncStopContinuousHandle = IntPtr.Zero;
        }

        internal void StartKeywordRecognition(KeywordRecognitionModel model)
        {
            if (asyncStartKeywordHandle != IntPtr.Zero)
            {
                ThrowIfFail(Internal.Recognizer.recognizer_async_handle_release(asyncStartKeywordHandle));
            }
            ThrowIfNull(recoHandle, "Invalid recognizer handle");
            ThrowIfFail(Internal.Recognizer.recognizer_start_keyword_recognition_async(recoHandle, model.keywordHandle, out asyncStartKeywordHandle));
            GC.KeepAlive(model);
            ThrowIfFail(Internal.Recognizer.recognizer_start_keyword_recognition_async_wait_for(asyncStartKeywordHandle, UInt32.MaxValue));
            ThrowIfFail(Internal.Recognizer.recognizer_async_handle_release(asyncStartKeywordHandle));
            asyncStartKeywordHandle = IntPtr.Zero;
        }

        internal void StopKeywordRecognition()
        {
            if (asyncStopKeywordHandle != IntPtr.Zero)
            {
                ThrowIfFail(Internal.Recognizer.recognizer_async_handle_release(asyncStopKeywordHandle));
            }
            ThrowIfNull(recoHandle, "Invalid recognizer handle");
            ThrowIfFail(Internal.Recognizer.recognizer_stop_keyword_recognition_async(recoHandle, out asyncStopKeywordHandle));
            ThrowIfFail(Internal.Recognizer.recognizer_stop_keyword_recognition_async_wait_for(asyncStopKeywordHandle, UInt32.MaxValue));
            ThrowIfFail(Internal.Recognizer.recognizer_async_handle_release(asyncStopKeywordHandle));
            asyncStopKeywordHandle = IntPtr.Zero;
        }

        internal delegate IntPtr GetRecognizerFromConfigDelegate(out IntPtr phreco, InteropSafeHandle speechconfig, InteropSafeHandle audioInput);

        internal static InteropSafeHandle FromConfig(GetRecognizerFromConfigDelegate fromConfig, SpeechConfig speechConfig, Audio.AudioConfig audioConfig)
        {
            if (speechConfig == null) throw new ArgumentNullException(nameof(speechConfig));
            if (audioConfig == null) throw new ArgumentNullException(nameof(audioConfig));

            IntPtr recoHandlePtr = IntPtr.Zero;
            ThrowIfFail(fromConfig(out recoHandlePtr, speechConfig.configHandle, audioConfig.configHandle));
            InteropSafeHandle recoHandle = new InteropSafeHandle(recoHandlePtr, Internal.Recognizer.recognizer_handle_release);
            GC.KeepAlive(speechConfig);
            GC.KeepAlive(audioConfig);
            return recoHandle;
        }

        internal static InteropSafeHandle FromConfig(GetRecognizerFromConfigDelegate fromConfig, SpeechConfig speechConfig)
        {
            if (speechConfig == null) throw new ArgumentNullException(nameof(speechConfig));

            IntPtr recoHandlePtr = IntPtr.Zero;
            IntPtr audioConfigPtr = IntPtr.Zero;
            InteropSafeHandle audioConfigHandle = new InteropSafeHandle(audioConfigPtr, null);
            ThrowIfFail(fromConfig(out recoHandlePtr, speechConfig.configHandle, audioConfigHandle));
            InteropSafeHandle recoHandle = new InteropSafeHandle(recoHandlePtr, Internal.Recognizer.recognizer_handle_release);
            GC.KeepAlive(speechConfig);
            return recoHandle;
        }
    }
}

File: source/bindings/java/com/microsoft/cognitiveservices/speech/Recognizer.java
//
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
//
package com.microsoft.cognitiveservices.speech;

import java.io.Closeable;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import com.microsoft.cognitiveservices.speech.audio.AudioConfig;
import com.microsoft.cognitiveservices.speech.util.EventHandlerImpl;
import com.microsoft.cognitiveservices.speech.util.IntRef;
import com.microsoft.cognitiveservices.speech.util.SafeHandle;
import com.microsoft.cognitiveservices.speech.util.SafeHandleType;
import com.microsoft.cognitiveservices.speech.util.Contracts;

/**
 * Defines the base class Recognizer which mainly contains common event handlers.
 */
public class Recognizer implements Closeable
{
    /*! \cond PROTECTED */

    protected static ExecutorService s_executorService;

    protected AtomicInteger eventCounter = new AtomicInteger(0);
    protected Integer backgroundAttempts = 0;
    /*! \endcond */

    static {
        s_executorService = Executors.newCachedThreadPool();
    }

    /**
     * Defines event handler for session started event.
     */
    final public EventHandlerImpl<SessionEventArgs> sessionStarted = new EventHandlerImpl<SessionEventArgs>(eventCounter);

    /**
     * Defines event handler for session stopped event.
     */
    final public EventHandlerImpl<SessionEventArgs> sessionStopped = new EventHandlerImpl<SessionEventArgs>(eventCounter);

    /**
     * Defines event handler for speech start detected event.
     */
    final public EventHandlerImpl<RecognitionEventArgs> speechStartDetected = new EventHandlerImpl<RecognitionEventArgs>(eventCounter);

    /**
     * Defines event handler for speech end detected event.
     */
    final public EventHandlerImpl<RecognitionEventArgs> speechEndDetected = new EventHandlerImpl<RecognitionEventArgs>(eventCounter);

    /*! \cond PROTECTED */

    /**
     * Creates and initializes an instance of a Recognizer
     * @param audioInput An optional audio input configuration associated with the recognizer
     */
    protected Recognizer(AudioConfig audioInput) {
        recoHandle = new SafeHandle(0, SafeHandleType.Recognizer);
        // Note: Since ais is optional, no test for null reference
        audioInputKeepAlive = audioInput;
    }

    /*! \endcond */

    /**
     * Dispose of associated resources.
     * Note: close() must be called in order to relinquish underlying resources held by the object.
     */
    @Override
    public void close() {
        synchronized (recognizerLock) {
            if (activeAsyncRecognitionCounter != 0) {
                throw new IllegalStateException("Cannot dispose a recognizer while async recognition is running. Await async recognitions to avoid unexpected disposals.");
            }
            dispose(true);
        }
    }

    /*! \cond INTERNAL */
    /**
     * Returns the internal recognizer instance
     * @return The internal recognizer instance
     */
    public SafeHandle getImpl() {
        return recoHandle;
    }
    /*! \endcond */
    
    /*! \cond PROTECTED */
    /**
     * This method performs cleanup of resources.
     * The Boolean parameter disposing indicates whether the method is called from Dispose (if disposing is true) or from the finalizer (if disposing is false).
     * Derived classes should override this method to dispose resource if needed.
     * @param disposing Flag to request disposal.
     */
    protected void dispose(final boolean disposing) {
        if (disposed) {
            return;
        }

        if (disposing) {
            if(this.eventCounter.get() != 0 && backgroundAttempts <= 50 ) {
                // There is an event callback in progress, closing while in an event call results in SWIG problems, so 
                // spin a thread to try again in 500ms and return.
                Thread t = new Thread(
                    new Runnable(){
                        @Override
                        public void run() {
                            try{
                                Thread.sleep(500 * ++backgroundAttempts);
                                dispose(disposing);
                            } catch (Exception e){}
                        }
                    });
                t.start();
            }
            else {
            }
        }
        audioInputKeepAlive = null;
        disposed = true;
    }

    protected void doAsyncRecognitionAction(Runnable recoImplAction) {
        synchronized (recognizerLock) {
            activeAsyncRecognitionCounter++;
        }
        if (disposed) {
            throw new IllegalStateException(this.getClass().getName());
        }
        try {
            recoImplAction.run();
        } finally {
            synchronized (recognizerLock) {
                activeAsyncRecognitionCounter--;
            }
        }
    }

    protected long recognize()
    {
        Contracts.throwIfNull(recoHandle, "Invalid recognizer handle");
        IntRef result = new IntRef(0);
        Contracts.throwIfFail(recognizeOnce(recoHandle, result));
        return result.getValue();
    }

    /*! \endcond */

    private void sessionStartedEventCallback(long eventArgs)
    {
        try {
            Contracts.throwIfNull(this, "recognizer");
            if (this.disposed) {
                return;
            }
            SessionEventArgs arg = new SessionEventArgs(eventArgs, true);
            EventHandlerImpl<SessionEventArgs> handler = this.sessionStarted;
            if (handler != null) {
                handler.fireEvent(this, arg);
            }    
        } catch (Exception e) {}
    }

    private void sessionStoppedEventCallback(long eventArgs)
    {
        try {
            Contracts.throwIfNull(this, "recognizer");
            if (this.disposed) {
                return;
            }
            SessionEventArgs arg = new SessionEventArgs(eventArgs, true);
            EventHandlerImpl<SessionEventArgs> handler = this.sessionStopped;
            if (handler != null) {
                handler.fireEvent(this, arg);
            }    
        } catch (Exception e) {}
    }

    private void speechStartDetectedEventCallback(long eventArgs)
    {
        try {
            Contracts.throwIfNull(this, "recognizer");
            if (this.disposed) {
                return;
            }
            RecognitionEventArgs arg = new RecognitionEventArgs(eventArgs, true);
            EventHandlerImpl<RecognitionEventArgs> handler = this.speechStartDetected;
            if (handler != null) {
                handler.fireEvent(this, arg);
            }    
        } catch (Exception e) {}
    }

    private void speechEndDetectedEventCallback(long eventArgs)
    {
        try {
            Contracts.throwIfNull(this, "recognizer");
            if (this.disposed) {
                return;
            }
            RecognitionEventArgs arg = new RecognitionEventArgs(eventArgs, true);
            EventHandlerImpl<RecognitionEventArgs> handler = this.speechEndDetected;
            if (handler != null) {
                handler.fireEvent(this, arg);
            }    
        } catch (Exception e) {}
    }

    /*! \cond PROTECTED */
    protected SafeHandle recoHandle = null;
    protected final native long startContinuousRecognition(SafeHandle recoHandle);
    protected final native long stopContinuousRecognition(SafeHandle recoHandle);
    protected final native long startKeywordRecognition(SafeHandle recoHandle, SafeHandle keywordModelHandle);
    protected final native long stopKeywordRecognition(SafeHandle recoHandle);
    protected final native long recognizingSetCallback(long recoHandle);
    protected final native long recognizedSetCallback(long recoHandle);
    protected final native long canceledSetCallback(long recoHandle);
    protected final native long sessionStartedSetCallback(long recoHandle);
    protected final native long sessionStoppedSetCallback(long recoHandle);
    protected final native long speechStartDetectedSetCallback(long recoHandle);
    protected final native long speechEndDetectedSetCallback(long recoHandle);
    protected final native long getPropertyBagFromRecognizerHandle(SafeHandle recoHandle, IntRef propertyHandle);
    /*! \endcond */

    private final native long recognizeOnce(SafeHandle recoHandle, IntRef resultHandle);

    private AudioConfig audioInputKeepAlive = null;
    private boolean disposed = false;
    private final Object recognizerLock = new Object();
    private int activeAsyncRecognitionCounter = 0;
}